cypress-io/cypress
# fix: improve keyboard behavior of the inline specs list "run all" buttons

- Closes #24762

### User facing changelog


The "Run N specs" directory-level buttons in the Inline Specs List now function correctly when activated by the keyboard.

### Additional details


This is an incremental improvement in the accessibility of this list so that keyboard behavior is better, even though it doesn't get us everything we'd like for accessibility here and there are further improvements that could be made (for example, nesting the lists to that the tree structure is better represented).

Nesting controls inside other controls should be avoided, but that is something we will need to come back to again as we revisit this experiment and the code around the specs lists more generally. I want to keep the diff small and easy to review.

The reason that keyboard events weren't working correctly previously is that the "directory item" had custom keyboard behavior to open/close a particular folder. Since it was not a button element, keydown event listeners had been added to it to simulate button behavior, and those listeners were capturing keyboard events firing on the internal button, before they had a chance to trigger the internal button's own `click` event.

This PR makes the toggle control for folders an actual button, which means keyboard events for space and enter are "free" since they trigger `click` events, so we can remove those listeners. Then the internal button functions correctly and things are overall a bit simpler.


### Steps to test


With `experimentalRunAllSpecs` turned on, open a spec and use the inline specs list. Arrow keys should move you up and down the list and close/open sections, tab key should let you reach the `run n specs` buttons. Enter and space keys have the following behaviors:

- Focused on a "directory" row, toggle it open and closed
- Focused on a "spec" row, run that spec
- Focused on a "run n specs" button, run the specs

On a project without `experimentalRunAllSpecs`, the list should still function as before.

The one change is that tab key will now also allow moving up and down items. Previously we had disabled this and required arrows for navigation, I think we could consider that again to make it easier to "skip" the list with the keyboard and access other content, but since these are actually interactive elements lets give that some more thought and find a way to communicate that arrow keys are the only way to navigate. As it stands I think it's more correct and predictable to let the browser handle tab order.

### How has the user experience changed?


Before: space and enter keys when used on the "Run N Specs" button in the Inline Specs List would just toggle the row

https://user-images.githubusercontent.com/8340719/205347610-d0d993b6-8aaa-4579-8ff5-147f91047dd7.mov


After: specs can be run with the keyboard

https://user-images.githubusercontent.com/8340719/205347660-7cfe0834-2617-4d17-8dbf-6e73d97fe47f.mov



### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
